src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs

```diff
@@ -420,31 +420,32 @@ private async Task StartAsyncCore(CancellationToken cancellationToken)
             CheckDisposed();
 
             Log.Starting(_logger);
-
-            // Start the connection
-            var connection = await _connectionFactory.ConnectAsync(_protocol.TransferFormat);
-            var startingConnectionState = new ConnectionState(connection, this);
-
-            // From here on, if an error occurs we need to shut down the connection because
-            // we still own it.
-            try
-            {
-                Log.HubProtocol(_logger, _protocol.Name, _protocol.Version);
-                await HandshakeAsync(startingConnectionState, cancellationToken);
-            }
-            catch (Exception ex)
+            using (var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _state.StopCts.Token))
             {
-                Log.ErrorStartingConnection(_logger, ex);
+                // Start the connection
+                var connection = await _connectionFactory.ConnectAsync(_protocol.TransferFormat, cts.Token);
+                var startingConnectionState = new ConnectionState(connection, this);
 
-                // Can't have any invocations to cancel, we're in the lock.
-                await CloseAsync(startingConnectionState.Connection);
-                throw;
-            }
+                // From here on, if an error occurs we need to shut down the connection because
+                // we still own it.
+                try
+                {
+                    Log.HubProtocol(_logger, _protocol.Name, _protocol.Version);
+                    await HandshakeAsync(startingConnectionState, cancellationToken);   //here we use cancellationToken, as HandshakeAsync combines cts for itself.
```

Comment:
Nit: All comments go on their own line.

This comment should no longer be necessary anyway after we change StartAsyncInner and ReconnectAsync to pass in a token linked to the StopCts and stop having HandshakeAsync do that itself.